Defra.Cdp.Backend.Api/Services/Deployments/DeploymentsService.cs

```diff
@@ -389,46 +390,75 @@ public async Task<Paginated<DeploymentOrMigration>> FindLatestWithMigrations(
                 break;
         }
 
-        var migrationCollection = Collection.Database.GetCollection<DatabaseMigration>("migrations");
-        var migrationPipeline = new EmptyPipelineDefinition<DatabaseMigration>()
-            .Match(migrationFilter)
-            .Project(m => new DeploymentOrMigration
-        {
-            Migration = m, Created = m.Created
-        });
-
-        var pipeline = new EmptyPipelineDefinition<Deployment>()
-            .Match(filter)
-            .Project(d => new DeploymentOrMigration { Deployment = d, Created = d.Created })
-            .UnionWith(migrationCollection, migrationPipeline)
-            .Sort(new SortDefinitionBuilder<DeploymentOrMigration>().Descending(d => d.Created))
-            .Skip(offset + size * (page - DefaultPage))
-            .Limit(size);
-        
-        var deployments = await Collection.Aggregate(pipeline).ToListAsync(ct);
-
-        // TODO favourites should be on the entire result - "bring all favourite things to the front", not just on the current page
-        //  we may be able to do this in an easier way in the UI
+        var favouriteServiceIds = new HashSet<string>();
         if (favourites?.Length > 0)
         {
             var repos = (await Task.WhenAll(favourites.Select(teamId =>
                     _repositoryService.FindRepositoriesByTeamId(teamId, true, ct))))
                 .SelectMany(r => r)
-                .ToList();
+                .Select(r => r.Id)
+                .ToHashSet();
 
-            var servicesOwnedByTeam = repos.Select(r => r.Id);
-            deployments = deployments.OrderByDescending(d =>
-                servicesOwnedByTeam.Contains(d.Deployment?.Service ?? d.Migration?.Service)
-            ).ToList();
+            favouriteServiceIds = repos;
         }
 
-        var totalDeployments = await Collection.CountDocumentsAsync(filter, cancellationToken: ct);
+        var favouriteArray = new BsonArray(favouriteServiceIds);
+        var deploymentStages = BuildAggregationStages(filter, "Deployment", favouriteArray);
+        var migrationStages = BuildAggregationStages(migrationFilter, "Migration", favouriteArray);
 
+        var pipeline = new List<BsonDocument>();
+        pipeline.AddRange(deploymentStages);
+        pipeline.Add(new BsonDocument("$unionWith",
+            new BsonDocument { { "coll", "migrations" }, { "pipeline", new BsonArray(migrationStages) } }));
+        pipeline.Add(new BsonDocument("$sort", new BsonDocument { { "isFavourite", -1 }, { "Created", -1 } }));
```

Comment:
Is this sort going to work? isFavourite isn't in the data (since it depends on who's asking)
Also, do we really need to do it all in raw bson objects? We're loosing the type safety for not a lot of gain

Reply:
yup all working as expected. Have added a demo ☝🏻

Reply:
> Also, do we really need to do it all in raw bson objects? We're loosing the type safety for not a lot of gain
If I have it correct, union and facet are not natively available. But I can have a refactor on some of this if you like?